keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -638,24 +640,107 @@ always dispatched immediately.  Following is how the other requests
+
+Each non-exempt PriorityLevelConfiguration's spec has an
+`assuredConcurrencyShares`, which has existed since APF was introduced
+and may not be zero, and a `lendableConcurrencyShares` field, which is
+being added in the midst of the lifetime of the `v1beta2` version of
+the API and may be any value between zero and
+`assuredConcurrencyShares` inclusive (default is zero).  Each
+apiserver allocates NominalConcurrencyLimits in proportion to
+`assuredConcurrencyShares` and LendableConcurrncyLimit in
+corresponding propotion:
+
+```
+NominalConcurrencyLimit(i)  = ceil( SCL * assuredConcurrencyShares(i)  / sum_assured )
+LendableConcurrencyLimit(i) = ceil( SCL * lendableConcurrencyShares(i) / sum_assured )
+NonLendableConcurrencyLimit(i) = NominalConcurrencyLimit(i) - LendableConcurrencyLimit(i)
+sum_assured  = sum[priority levels k] assuredConcurrencyShares(k)
+```
+
+where SCL is the apiserver's concurrency limit.
+
+Borrowing is further limited by a practical consideration: we do not
+want a global mutex covering all dispatching.  Aside from borrowing,
+dispatching from one priority level is done independently from
+dispatching at another.  Borrowing is allowed in just one direction
+(higher may borrow from lower, and lower does not actively lend to
+higher) and in a very limited quantity: an attempt to dispatch for one
+priority level will consider borrowing from just one other priority
+level (if there is any lower priority level at all).
+
+A request can be dispatched exactly at a non-exempt priority level
+when either there are no requests executing at that priority level or
+the number of seats needed by that request is no greater than the
+number of unused seats at that priority level.  The number of unused
+seats at a given priority level is that level's
+NominalConcurrencyLimit minus the number of seats used by requests
+executing at that priority level (dispatched from that priority level
+and higher ones).
+
+There are two sorts of times when dispatching to a non-empty priority
+level is considered: when a request arrives, and when a request
+releases the seats it was occupying (which is not the same as when the
+request finishes from the client's point of view, see below about
+WATCH requests).
+
+At each of these sorts of moments, as many requests are dispatched
+exactly at the same priority level as possible.  The next request to
+consider dispatching is chosen by using the Fair Queuing for Server
+Requests algorithm below to choose a queue at that priority level, and
+the request at the head of that queue is considered.  If (a) no
+requests can be dispatched exactly at that priority level at that
+moment, (b) there are non-empty queues at that level, and (c) there
+are lower non-exempt priority levels, then the request at the head of
+the chosen queue is considered for dispatch at one of the lower
+priority levels.  The particular lower priority level considered is
+drawn at random from the lower ones, in proportion to their
```

Comment:
I'm not a fan of this. And thinking about it, I'm wondering if we really need that.

The critical property you're introducing is effectively:

   "A given request can borrow from at most one other priority level"

That is a nice simplification and I think I can buy its consequences of potentially not being able to dispatch a request even if in theory it would be dispatchable if we would be able to borrow from multiple PLs. Because the additional cognitive benefit from that simplification is fairly huge.


Now - I don't really buy the fact that we will be choosing the PL to borrow from at random. Any randomization here will make it fairly hard to understand/debug what was really happening in my production cluster. And I don't think I really see why we need that.

Here is how I think about it:
1. we have a set of (ordered) PLs [the set itself is protected by its own lock, but individual QueueSet per PL are protected on their own by their own locks]
2. when we can't dispatch a request R for PL (due to lack of seats) then we:
- grab a reader lock from (1) [this seems cheap, because PLs are added/removed really rarely, so that's fine]
- iterate over all PLs with lower `priority` and for each of them:
  (a) take a lock on that PL
  (b) check if it has enough seats so that we can borrow to satisfy our request
  (c) if so - borrow and return
  (d) release the lock from that PL

In the above approach, we don't really need a global lock - we still lock exactly one PL at a time, which seems what we need.

Reply:
Based on our discussion in the meeting today, I changed the proposal to have the higher priority level sequentially consider borrowing from each lower priority level.  And I also added an explicit statement that all the seats are taken from one priority level.  I also added some words about how little the existing global lock is used.